_posts/2017-06-21-Devoxx-pl.md

```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
 
 Extended blog coming later.
 
+## Feature Branches And Toggles In A Post-GitHub World ([Sam Newman](https://twitter.com/@samnewman))
+
+Sam told us about his experience at a project that was having trouble merging branches.
+The release branch for the next release was called R3, but for a large refactoring, branch R4 was created.
+Afterwards he described merging the branches as a car crash.
+They even needed to introduce a dedicated R3-R4 merge bug fix team.
+Then **Continuous Integration** was meant to prevent the merging problem.
+Developers had to check in their changes and the CI tool would validate the code.
+The problem with R3-R4 release was that validation was done only for a branch, not on the integrated branches.
+
+>The integration should be validated every day and when the build breaks, fix it!
+
+For unfinished work we can wait till it is ready before checking in.
+This exposes us to the risk of losing work when it's only on the developers computer.
+
+### Feature Branch
+
+An alternative would be to create a **feature branch**, which brings us back to the problem of merging branches.
+
+>Pain of merge = fn(size_of_merge, duration_since_last_merge)
+
+Merging branches can be a difficult task and might lead to commit race, to offload the effort to a colleague.
+
+### Trunk-based development
+
+A third option would be to 'check in anyway', called **trunk-based development**.
+Every commit integrates to the trunk and developers should integrate their local changes daily.
+Small changes and integrating often makes it easier to merge new code.
+
+New half-finished features can be hidden with **feature toggles**.
+These toggles can me managed using flags or configurations (eg, in Zookeeper, Consul, ...).
+>A flag should be set and evaluated in as few places as possible, preferably once each.
+>Flags should be removed when the new implementation is done.
+
+More info: [Trunk-based development](https://trunkbaseddevelopment.com/)
+
+Changes to existing functionality can be done by providing an abstraction above the existing functionality.
+The new functionality can then be developed for the abstraction and when it is done, change to the new implementation.
+**Branch-by abstraction** has the side-benefit that it can be used for **A/B and canary releasing**.
+
+[The Continuous Delivery](https://www.amazon.de/Continuous-Delivery-Deployment-Automation-Addison-Wesley/dp/0321601912) book tells us to treat every check in as a possible release candidate.
+Developers start with the assumption it is worthy, CI tool decides if it is.
+Deploy frequently with small changes, makes easier to rollback, lowering the chance of problems.
+
+### Git
+
+And then there was **Git**, developed by [Linus Torvalds](https://en.wikipedia.org/wiki/Linus_Torvalds) with the goal to merge a patch in less than three seconds.
+In Git branches where much more lightweight and every repository contains the full source history.
```

Comment:
I assume you mean local repository?
In Git, branches **are** much more lightweight and every **local** repository contains the full source history.